From db42a0f0c5014790f2e4ebffa9a16e5f6d78e492 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Wed, 19 Oct 2022 11:57:59 +0200
Subject: [PATCH 0790/1010] evl/wait: lockdep: introduce
 evl_init_wait_on_stack()

We need a static lock class for wait queues we want to initialize from
oob context, we cannot register classes dynamically since this would
involve in-band only code, such as RCU read sections.

We may assume that only stack-based wait queues may have to be
initialized dynamically from the oob stage. Let's provide
evl_init_wait_on_stack() for this purpose, which defines a static lock
class key instead.

Extend the flag API which is based on wait queues to provide
evl_init_flag_on_stack() as well.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/flag.h |  6 ++++++
 include/evl/wait.h | 23 +++++++++++++++++++++--
 kernel/evl/wait.c  | 40 ++++++++++++++++++++++++++--------------
 3 files changed, 53 insertions(+), 16 deletions(-)

diff --git a/include/evl/flag.h b/include/evl/flag.h
index 44633dd04f..08c2237316 100644
--- a/include/evl/flag.h
+++ b/include/evl/flag.h
@@ -29,6 +29,12 @@ static inline void evl_init_flag(struct evl_flag *wf)
 	wf->raised = false;
 }
 
+static inline void evl_init_flag_on_stack(struct evl_flag *wf)
+{
+	evl_init_wait_on_stack(&wf->wait, &evl_mono_clock, EVL_WAIT_PRIO);
+	wf->raised = false;
+}
+
 static inline void evl_destroy_flag(struct evl_flag *wf)
 {
 	evl_destroy_wait(&wf->wait);
diff --git a/include/evl/wait.h b/include/evl/wait.h
index 429b83bc1d..354e55c807 100644
--- a/include/evl/wait.h
+++ b/include/evl/wait.h
@@ -31,15 +31,27 @@ struct evl_wait_channel {
 	const char *name;
 };
 
+#ifdef CONFIG_PROVE_LOCKING
+struct evl_lock_key_addr {
+	struct lock_class_key *addr;
+};
+#define __EVL_LOCK_KEY_ADDR_INITIALIZER  (struct foo){ .addr = NULL }
+#else
+struct evl_lock_key_addr { };
+#define __EVL_LOCK_KEY_ADDR_INITIALIZER  (struct foo){ }
+#endif
+
 struct evl_wait_queue {
 	int flags;
 	struct evl_clock *clock;
 	struct evl_wait_channel wchan;
+	struct evl_lock_key_addr lock_key_addr;
 };
 
 #define EVL_WAIT_INITIALIZER(__name) {					\
 		.flags = EVL_WAIT_PRIO,					\
 		.clock = &evl_mono_clock,				\
+		.lock_key_addr = __EVL_LOCK_KEY_ADDR_INITIALIZER,	\
 		.wchan = {						\
 			.lock = __HARD_SPIN_LOCK_INITIALIZER((__name).wchan.lock), \
 			.pi_serial = 0,					\
@@ -108,10 +120,17 @@ static inline bool evl_wait_active(struct evl_wait_queue *wq)
 void __evl_init_wait(struct evl_wait_queue *wq,
 		struct evl_clock *clock,
 		int flags,
-		const char *name);
+		const char *name,
+		struct lock_class_key *lock_key);
 
 #define evl_init_wait(__wq, __clock, __flags)	\
-	__evl_init_wait(__wq, __clock, __flags, #__wq)
+	__evl_init_wait(__wq, __clock, __flags, #__wq, NULL)
+
+#define evl_init_wait_on_stack(__wq, __clock, __flags)	\
+	do {								\
+		static struct lock_class_key __key;			\
+		__evl_init_wait(__wq, __clock, __flags, #__wq, &__key); \
+	} while (0)
 
 void evl_destroy_wait(struct evl_wait_queue *wq);
 
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index 9fc2c11080..c2c161a57c 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -14,7 +14,8 @@
 
 void __evl_init_wait(struct evl_wait_queue *wq,
 		struct evl_clock *clock, int wq_flags,
-		const char *name)
+		const char *name,
+		struct lock_class_key *lock_key)
 {
 	unsigned long flags __maybe_unused;
 
@@ -28,18 +29,25 @@ void __evl_init_wait(struct evl_wait_queue *wq,
 	INIT_LIST_HEAD(&wq->wchan.wait_list);
 	raw_spin_lock_init(&wq->wchan.lock);
 #ifdef CONFIG_PROVE_LOCKING
-	lockdep_register_key(&wq->wchan.lock_key);
-	lockdep_set_class_and_name(&wq->wchan.lock, &wq->wchan.lock_key, name);
-	/*
-	 * might_lock() forces lockdep to pre-register a dynamic lock
-	 * class, instead of waiting lazily for the first lock
-	 * acquisition, which might happen oob for us. Since that
-	 * registration depends on RCU, we need to make sure this
-	 * happens in-band.
-	 */
-	local_irq_save(flags);
-	might_lock(&wq->wchan.lock);
-	local_irq_restore(flags);
+	if (lock_key) {
+		wq->lock_key_addr.addr = lock_key;
+		lockdep_set_class_and_name(&wq->wchan.lock, lock_key, name);
+	} else {
+		lock_key = &wq->wchan.lock_key;
+		wq->lock_key_addr.addr = lock_key;
+		lockdep_register_key(lock_key);
+		lockdep_set_class_and_name(&wq->wchan.lock, lock_key, name);
+		/*
+		 * might_lock() forces lockdep to pre-register a
+		 * dynamic lock class, instead of waiting lazily for
+		 * the first lock acquisition, which might happen oob
+		 * for us. Since that registration depends on RCU, we
+		 * need to make sure this happens in-band.
+		 */
+		local_irq_save(flags);
+		might_lock(&wq->wchan.lock);
+		local_irq_restore(flags);
+	}
 #endif
 }
 EXPORT_SYMBOL_GPL(__evl_init_wait);
@@ -48,7 +56,11 @@ void evl_destroy_wait(struct evl_wait_queue *wq)
 {
 	evl_flush_wait(wq, T_RMID);
 	evl_schedule();
-	lockdep_unregister_key(&wq->wchan.lock_key);
+#ifdef CONFIG_PROVE_LOCKING
+	/* Drop dynamic key. */
+	if (wq->lock_key_addr.addr == &wq->wchan.lock_key)
+		lockdep_unregister_key(&wq->wchan.lock_key);
+#endif
 }
 EXPORT_SYMBOL_GPL(evl_destroy_wait);
 
-- 
2.30.2

